rp2: machine_uart.c: Enable buffered transfer of data.

robert-hh opened the pull request:
This PR adds support buffered transfer of data. The buffer size
is set using  rxbuf and txbuf keywords during instantiation
or init. The default and minimal size is 256. The maximal
size is 32766 (2**15 - 2).

uart.write() still includes checks for timeout, even if it is very
unlikely to happen due to a) lack of flow control support and
b) the minimal timeout values being longer than the time it needs
to send a byte .

robert-hh committed:
machine_uart.c: Add buffered transfer of data.

Instantiation and init now support the rxbuf and txbuf keywords for
setting the buffer size. The default and minimal size is 256.
The maximal size is 32766.

uart.write() still includes checks for timeout, even if it is very
unlikely to happen due to a) lack of flow control support and
b) the minimal timeout values being longer than the time it needs
to send a byte .

dpgeorge commented:
You accidentally (?) updated/downgraded the pico-sdk submodule.

robert-hh committed:
rp2: machine_uart.c: Update pico-sdk.

Hope it sticks.

robert-hh commented:
Definitely accidentally. There is always trouble when the submodule changes. submodule deinit lib/pico-sdk and again submodule update --init brings always back the old version.

robert-hh committed:
rp2: machine_uart.c: Fix the code formatting.

Due to complaints of the code formatter check.

robert-hh committed:
rp2:machine_uart.c: Simplify buffer allocation.

Do not try to deallocate previously use buffers. That led to problems
after warm boot. Just let GC to the job.

robert-hh committed:
machine_uart.c: Update the ioctl function too.

ioctl() also has to use the ring buffer functions.

robert-hh committed:
machine_uart.c: Introduce MIN_BUFFER_SIZE.

Which is different and smaller than DEFAULT_BUFFER SIZE.

kevinkk525 commented:
Any reason this PR can't get merged yet?

Hermann-SW commented:
> Any reason this PR can't get merged yet?

Not that I know of, I am waiting for this to get merged as well.
I use the bigger rxbuf to allow the PRaspberry Pi Pico doing https requests against the internet via serially connected ESP01S module's Wifi, see animation in this posting:
https://forum.micropython.org/viewtopic.php?f=2&p=56283#p56290

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -169,6 +252,36 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
+        // minimal size is DEFAULT_BUFFER_SIZE
+        // maximal size is MAX_BUFFER_SIZE
+        if (rxbuf_len < DEFAULT_BUFFER_SIZE) {
+            rxbuf_len = DEFAULT_BUFFER_SIZE;
```
Is it needed to protect against small buffers?  What's wrong with a small buffer like 4 bytes?

robert-hh replied:
I had a version with a MIN_BUFFER_SIZE. I'll enable that. The smallest reasonbale size seems to the the FIFO size, which is 32.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -169,6 +252,36 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
+        // minimal size is DEFAULT_BUFFER_SIZE
+        // maximal size is MAX_BUFFER_SIZE
+        if (txbuf_len < DEFAULT_BUFFER_SIZE) {
+            txbuf_len = DEFAULT_BUFFER_SIZE;
```
As above, I think the check for minimum can be much less, like 2.

robert-hh replied:
As above.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -192,14 +305,28 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
+        MP_STATE_PORT(rp2_uart_tx_buffer[uart_id]) = self->write_buffer.buf;
+
+        // set the irq handler
+        irq_set_exclusive_handler(self->irq_nr, self->irq_handler);
```
This could be:
```
if (self->uart_id == 0) {
    irq_set_exclusive_handler(UART0_IRQ, uart0_irq_handler);
} else {
    irq_set_exclusive_handler(UART1_IRQ, uart1_irq_handler);
}
```
which would save some RAM not having to have these variables in the instance.  (Alternatively have a separate static table with irq number and handler.)

robert-hh replied:
That looks good and saves RAM.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -192,14 +305,28 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
     machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
-    return MP_OBJ_NEW_SMALL_INT(uart_is_readable(self->uart));
+    // get all bytes from the fifo first
+    uart_drain_rx_fifo(self, true);
```
I don't think there needs to be any locking here.  The ringbuf is implemented so that it can be simultaneously read and written without any race conditions.

robert-hh replied:
I was not sure about the locking. I added that since during testing I had error during reading from the uart while data was sent to it. These errors went ways after adding the lock. The decision of full vs empty is done by comparing the read and write index. So reading and writing are not independent.

robert-hh replied:
When I disable the locking, the errors are back in the receive test.

robert-hh replied:
The concurrency is not between read and write of the ringbuf, but between two instances writing to the rinfbuf, a) the ISR, b) the polling loop in uart_read() and the call in uart_any(). The latter ones are needed because the ISR is not called until 4 items are in the FIFO.

dpgeorge replied:
Yes, agreed.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -246,8 +373,10 @@ STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t siz
             }
             MICROPY_EVENT_POLL_HOOK
+            // force few incoming bytes to the buffer
+            uart_drain_rx_fifo(self, true);
```
Is this needed?  I would think that the UART IRQ would trigger and update the ringbuf if there were any characters (and then locking is not needed).

robert-hh replied:
The irq is triggered on transiting a certain level in the FIFO, which can be configured is set to 4 during init as the minimal level. Less than 4 characters will not trigger the IRQ.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -272,7 +410,8 @@ STATIC mp_uint_t machine_uart_write(mp_obj_t self_in, const void *buf_in, mp_uin
             MICROPY_EVENT_POLL_HOOK
         }
-        uart_get_hw(self->uart)->dr = *src++;
+        ringbuf_put(&(self->write_buffer), *src++);
```
I think there's a (rare) race condition here:
- ringbuf is full so while loop above is waiting for an empty slot
- some long-running callback is executed during `MICROPY_EVENT_POLL_HOOK`
- UART TX IRQ drains the ring buffer
- UART TX IRQ has nothing left to send so no more TX IRQs are generated
- while loop above finishes
- ringbuf_put on this line adds more to the ringbuf
- data does not get put out on UART because there are no more TX IRQs

robert-hh replied:
That's indeed a clever observation. I can add another call to 'uart_fill_tx_fifo(self, true)`, which will start sending the remaining characters. If the FIFO is already full, it will do nothing.

dpgeorge replied:
I think the issue is still there.  The `uart_fill_tx_fifo()` call needs to be within this loop, otherwise `while (ringbuf_free(...)) { ... }` will time out if it hits this scenario.

For efficiency, probably best to put `uart_fill_tx_fifo()` within that wait-ringbuf-free loop.

robert-hh replied:
Right. I faintly recall having had it there once.

robert-hh replied:
Updated. Meanwhile I have a working (and failing) test case showing the need for the write lock in addition to the test for the read lock.

dpgeorge replied:
Ok, that's great you can test these subtle conditions.

Does it make sense to put these tests in to this repo, somehow?  If they are generic `machine.UART` tests then they could be used to test any port's UART.

At least, I would like to run the tests, if it's possible to share what you did?

robert-hh replied:
Sure. It is a little script with both a recv and send call. I use two RP2's for that test. Script below. It's a little bit hacky. It sends blocks of '0123456789', and the receiver tests for consistency. You run recv() on one board and send() on the other. As soon as you disable the read lock, you will get recv errors. It takes a while. They are not very frequent. When you disable the write lock, you will also get receive errors, but that more frequent.
```
from machine import UART
import time

def send(baud=460800):

    u = UART(0, baud, tx=Pin(12), rx=Pin(13), txbuf=1024, timeout_char = 1000, timeout=1000)
    # u = UART(1, baud, tx=Pin(8), rx=Pin(9), txbuf=200, timeout_char = 1000, timeout=1000)
    m="0123456789"
    sum = 0
    count=0
    try:
        while True:
            for i in range(5):
                sum += u.write(m*20)
                time.sleep_ms(2)
            print(sum)
            time.sleep(1)
            count += 1
    except KeyboardInterrupt:
        pass

def recv(baud=460800):
    u = UART(0, baud, tx=Pin(12), rx=Pin(13), rxbuf=1024)
    # u = UART(1, baud, tx=Pin(8), rx=Pin(9), rxbuf=1024)

    u.read(u.any())
    sum = 0
    try:
        nullerror = 0
        seqerror = 0
        enderror = 0
        while True:
            new = u.any()
            if new:
                while new:
                    sum += new
                    msg = u.read(new)
                    print(msg)
                    start = msg[0]
                    for c in msg[1:]:
                        if ((start + 1) != c) and not (start == 57 and c == 48):
                            if c != 0:
                                seqerror += 1
                            else:
                                nullerror += 1
                        start = c
                    time.sleep_ms(1)
                    new = u.any()
                if msg[-1] != 57:  # must be 9
                    enderror += 1
                print(msg, "\n", sum, seqerror, nullerror, enderror)

    except KeyboardInterrupt:
        if u.any():
            print(u.read(u.any()))
    print(sum)
```

dpgeorge replied:
Thanks!  I ran this code for about 13Mbytes of transfer and saw no problems.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -284,10 +423,10 @@ STATIC mp_uint_t machine_uart_ioctl(mp_obj_t self_in, mp_uint_t request, mp_uint
         uintptr_t flags = arg;
         ret = 0;
-        if ((flags & MP_STREAM_POLL_RD) && uart_is_readable(self->uart)) {
+        if ((flags & MP_STREAM_POLL_RD) && (ringbuf_avail(&self->read_buffer) > 0)) {
```
minor point: parenthesis are not needed around this `> 0` condition

robert-hh replied:
Removed

dpgeorge commented:
I don't think locks are needed at all.  Eg on stm32 there is no locking of the ringbuf for UART RX.

robert-hh committed:
machine_uart.c: Set of small changes suggested by Damien.

- introduce a MIN_BUFFER_SIZE, here set to 32
- remove irq_adress and irq_nr from the uart data type, saving ram
- add an additional tx_fifo_fill after putting the remaining bytes
  into the TX ringbuf.
- remove obsolete brackets

robert-hh commented:
I disabled the locking, did my receive burst test, and got receive errors. Not many, about one every 150k. But more than with locking, where I had no error in transmitting about one gigabyte of data.

robert-hh commented:
Disabling locking on send seems to be OK. I could not create a test case which had faults in sending.

robert-hh committed:
machine_uart.c: Move handling of lock flags.

Setting and releasing the flags is moved into the API functions instaed
of just passing an argument. That shrinks the ISR code.
The write-lock is still present, event if I could not yet create a test
case of failing without the lock. The check for timeout on write is
also still included, even if it cannot happen due to the absence of
a transmit flow control.

robert-hh committed:
machine_uart.c: Move the fifo fill call.

Avoiding a possible stuck in sending.

dpgeorge commented:
Squashed and merged in 1be74b94b6f3263b8e360a0a012ae87301539f91

robert-hh commented:
In response to a thread in the forum (https://forum.micropython.org/viewtopic.php?f=21&t=10299&start=10#p57109) I tested fast bursts of messages with a txbuf setting of 32. I have seen two events:
a) Timeout errors. These happen, if the message cannot be placed into the transfer buffer at once. Obviously, setting a timeout value cures that. But I would not have expected a timeout in that situation. So the question is not programmatic, but more semantic: If a use should expect a timeout at all in that situation.
b) In combination with threading various kinds of severe error, from lock-up, doubling characters on transmisston, completely messed up message contents, ...  Even if I see it during uart.write(), it seems not to be related to it. It seems more a interference between threading and Interrupts. The poster Tinus seemed to have that problem already before. I'll open a separate Issue about it.

QGB commented:
@robert-hh micropython uart  uasyncio

robert-hh commented:
@QGB Yes? Why do you ring the bell?
